types/react/index.d.ts

```diff
@@ -1081,7 +1081,7 @@ declare namespace React {
     // A specific function type would not trigger implicit any.
     // See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/52873#issuecomment-845806435 for a comparison between `Function` and more specific types.
     // tslint:disable-next-line ban-types
-    function useCallback<T extends Function>(callback: T, deps: DependencyList): T;
+    function useCallback<T extends Function | null | undefined>(callback: NonNullable<T>, deps: DependencyList): NonNullable<T>;
```

Comment:
I think I'm ok with this convenience constraint. But the problem will continue if you pass it to something like `Function | number` where it'll also fail to infer the event type. 

Though I wonder how often this really happens in practice. Because the test given (assuming we move it out of `useLayoutEffect` would break the rules of hooks very easily since you wouldn't be able to write it that way in a loop or in a render with multiple returns.

 So I'm not sure we should allow this and not force people to write out the parameter type. 

I'll rewrite the test to use idiomatic React (no rules-of-hook violation) and then add some to highlight the problem. Maybe we can find the solution that works for all.